drivers/usbc/tcpc/ucpd_stm32.c

```diff
@@ -0,0 +1,1463 @@
+	 */
+
+	if (info->evt & UCPD_EVT_GOOD_CRC_REQ) {
+		data->ucpd_tx_request |= MSG_GOOD_CRC_MASK;
+	}
+
+	if (info->evt & UCPD_EVT_TCPM_MSG_REQ) {
+		data->ucpd_tx_request |= MSG_TCPM_MASK;
+	}
+
+	/*
+	 * Manage PD tx messages. The state machine may need to be
+	 * called more than once. For instance, if
+	 * the task is woken at the completion of sending a GoodCRC,
+	 * there may be a TCPM message request pending and just changing
+	 * the state back to idle would not trigger start of transmit.
+	 */
+	do {
+		ucpd_manage_tx(info);
+		/* Look at events only once. */
+		info->evt = 0;
+	} while (data->ucpd_tx_request &&
+		 data->ucpd_tx_state == STATE_IDLE &&
+		 !data->ucpd_rx_msg_active);
+}
+
+/**
+ * @brief Sends a goodCRC message
+ */
+static void ucpd_send_good_crc(const struct device *dev,
+			       union pd_header rx_header)
+{
+	struct tcpc_data *data = dev->data;
+	const struct tcpc_config *const config = dev->config;
+	union pd_header tx_header;
+	enum pd_packet_type tx_type;
+	struct alert_info *info = &data->alert_info;
+
+	/*
+	 * A GoodCRC message shall be sent by receiver to ack that the previous
+	 * message was correctly received. The GoodCRC message shall return the
+	 * rx message's msg_id field. The one exception is for GoodCRC messages,
+	 * which do not generate a GoodCRC response
+	 */
+	if (ucpd_msg_is_good_crc(rx_header)) {
+		return;
+	}
+
+	/*
+	 * Get the rx ordered set code just detected. SOP -> SOP''_Debug are in
+	 * the same order as enum tcpc_packet_type and so can be used
+	 * directly.
+	 */
+	tx_type = LL_UCPD_ReadRxOrderSet(config->ucpd_port);
+
+	/*
+	 * PD Header(SOP):
+	 *   Extended   b15    -> set to 0 for control messages
+	 *   Count      b14:12 -> number of 32 bit data objects = 0 for ctrl msg
+	 *   MsgID      b11:9  -> running byte counter (extracted from rx msg)
+	 *   Power Role b8     -> stored in static, from set_msg_header()
+	 *   Spec Rev   b7:b6  -> PD spec revision (extracted from rx msg)
+	 *   Data Role  b5     -> stored in static, from set_msg_header
+	 *   Msg Type   b4:b0  -> data or ctrl type = PD_CTRL_GOOD_CRC
+	 */
+	/* construct header message */
+	tx_header.message_type = PD_CTRL_GOOD_CRC;
+	if (tx_type == PD_PACKET_SOP) {
+		tx_header.port_power_role = data->msg_header.pr;
+		tx_header.port_data_role = data->msg_header.dr;
+	} else {
+		tx_header.port_power_role = 0;
+		tx_header.port_data_role = 0;
+	}
+	tx_header.message_id = rx_header.message_id;
+	tx_header.number_of_data_objects = 0;
+	tx_header.specification_revision = rx_header.specification_revision;
+	tx_header.extended = 0;
+
+	/* Good CRC is header with no other objects */
+	data->ucpd_tx_buffers[TX_MSG_GOOD_CRC].msg_len = 2;
+	data->ucpd_tx_buffers[TX_MSG_GOOD_CRC].data.header =
+		tx_header.raw_value;
+	data->ucpd_tx_buffers[TX_MSG_GOOD_CRC].type = tx_type;
+
+	/* Notify ucpd task that a GoodCRC message tx request is pending */
+	info->evt |= UCPD_EVT_GOOD_CRC_REQ;
```

Comment:
offloading good crc to worker will add undefined delay ( based on pending work item in the work queue).
how can the gooc-crc timing of ~1msec be achieved all the time?

Reply:
Same comment above